src/cycle_basis.c

```diff
@@ -0,0 +1,660 @@
+/* -*- mode: C -*-  */
+/* 
+   IGraph library.
+   Copyright (C) 2010-2012  Gabor Csardi <csardi.gabor@gmail.com>
+   334 Harvard street, Cambridge, MA 02139 USA
+   Copyright (C) 2019 the igraph team.
+   
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 
+   02110-1301 USA
+
+*/
+
+#include "igraph_structural.h"
+#include "igraph_dqueue.h"
+#include "igraph_interface.h"
+#include "igraph_interrupt_internal.h"
+#include "igraph_memory.h"
+#include "igraph_types_internal.h"
+#include "igraph_visitor.h"
+
+/**
+ * \function igraph_fundamental_cycle_basis
+ * \brief Find a fundamental cycle basis for an unweighted, undirected graph
+ * 
+ * A cycle basis is a set of cycles such that the set difference
+ * between them (i.e. the set of edges that are counted an odd
+ * number of times by the cycles) is a Eulerian path across the
+ * graph, which means it touches all vertices once and never
+ * walks the same edge more than once.
+ *
+ * A fundamental cycle basis is found by calculating a spanning tree
+ * and closing each pair of nonadjacent vertices with an external edge
+ * if that edge is present in the graph.
+ *
+ * \param graph The input graph, it might be directed, but edge
+ *    direction is ignored.
+ * \param basis A pointer to a vector of vectors. Each fundamental cycle
+ *    will be a member as a list of edge ids.
+ * \return Error code.
+ * 
+ * Time complexity: O(|V| + |E|) (spanning tree) + ?? (backtracking).
+ * 
+ */
+int igraph_fundamental_cycle_basis(const igraph_t *graph,
+    igraph_vector_ptr_t *basis) {
```

Comment:
Don't forget that the user may pass in a non-empty `vector_ptr`. Clear the `vector_ptr` when you receive it (an analogous issue had to be corrected in Yasir's PR as well).